File: app/src/main/java/com/commit451/gitlab/tools/ConversionUtil.java
package com.commit451.gitlab.tools;

import com.commit451.gitlab.api.GitLabClient;
import com.commit451.gitlab.model.Account;

import org.joda.time.DateTime;
import org.joda.time.format.ISODateTimeFormat;

import android.net.Uri;

import java.util.Date;

public final class ConversionUtil {
    private ConversionUtil() {}

    public static String fromDate(Date date) {
        return ISODateTimeFormat.dateTime().print(new DateTime(date));
    }

    public static Date toDate(String dateString) {
        if (dateString == null || dateString.isEmpty()) {
            return null;
        }

        return ISODateTimeFormat.dateTimeParser().parseDateTime(dateString).toDate();
    }

    public static String fromUri(Uri uri) {
        if (uri == null) {
            return null;
        }

        return uri.toString();
    }

    public static Uri toUri(String uriString) {
        if (uriString == null) {
            return null;
        }
        if (uriString.isEmpty()) {
            return Uri.EMPTY;
        }

        Uri uri = Uri.parse(uriString);
        if (!uri.isRelative()) {
            return uri;
        }

        Account account = GitLabClient.getAccount();
        if (account == null) {
            return uri;
        }

        Uri.Builder builder = Uri.parse(account.getServerUrl())
                .buildUpon()
                .encodedQuery(uri.getEncodedQuery())
                .encodedFragment(uri.getEncodedFragment());

        if (uri.getPath().startsWith("/")) {
            builder.encodedPath(uri.getEncodedPath());
        } else {
            builder.appendEncodedPath(uri.getEncodedPath());
        }

        return builder.build();
    }
}

File: app/src/main/java/com/commit451/gitlab/util/ConversionUtil.kt
package com.commit451.gitlab.util

import android.net.Uri
import com.commit451.gitlab.model.Account

/**
 * Converts things!
 */
object ConversionUtil {

    fun fromUri(uri: Uri?): String? {
        if (uri == null) {
            return null
        }

        return uri.toString()
    }

    fun toUri(account: Account?, uriString: String?): Uri? {
        if (uriString == null) {
            return null
        }
        if (uriString.isEmpty()) {
            return Uri.EMPTY
        }

        val uri = Uri.parse(uriString)
        if (!uri.isRelative) {
            return uri
        }

        if (account == null) {
            return uri
        }

        val serverUri = Uri.parse(account.serverUrl)
        val builder = serverUri
                .buildUpon()
                .encodedQuery(uri.encodedQuery)
                .encodedFragment(uri.encodedFragment)

        if (uri.path.startsWith("/")) {
            builder.encodedPath(uri.encodedPath)
        } else {
            builder.appendEncodedPath(uri.encodedPath)
        }

        return builder.build()
    }
}
